dkg/actions_signing.go

```diff
@@ -0,0 +1,84 @@
+package dkg
+
+import (
+	"errors"
+
+	"github.com/drand/drand/util"
+
+	"github.com/drand/drand/key"
+	"github.com/drand/drand/protobuf/drand"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+func (d *DKGProcess) signMessage(beaconID, messageType string, proposal *drand.ProposalTerms) (*drand.DKGMetadata, error) {
+	kp, err := d.beaconIdentifier.KeypairFor(beaconID)
+	if err != nil {
+		return nil, err
+	}
+
+	sig, err := kp.Scheme().AuthScheme.Sign(kp.Key, messageForProto(proposal, messageType, beaconID))
+	if err != nil {
+		return nil, err
+	}
+	return &drand.DKGMetadata{
+		BeaconID:  beaconID,
+		Address:   kp.Public.Address(),
+		Signature: sig,
+	}, nil
+}
+
+func (d *DKGProcess) verifyMessage(messageType string, metadata *drand.DKGMetadata, proposal *drand.ProposalTerms) error {
+	participants := util.Concat(proposal.Remaining, proposal.Joining)
+	// signing is done before the metadata is attached, so we must remove it before we perform verification
+	proposal.Metadata = nil
+
+	// find the participant the signature is allegedly from
+	var p *drand.Participant
+	for _, participant := range participants {
+		if participant.Address == metadata.Address {
+			p = participant
+			break
+		}
+	}
+	if p == nil {
+		return errors.New("no such participant")
+	}
+
+	// get the scheme for the network so we can correctly unmarshal the public key
+	kp, err := d.beaconIdentifier.KeypairFor(metadata.BeaconID)
+	if err != nil {
+		return err
+	}
+
+	// use that scheme to verify the message came from the alleged author
```

Comment:
Mhhhh I'm not sure i understand: you are reading the public key alongside the signature from the packet ? But then what is the point of signing if you don't know _who_ is signing ? Anyone can create a packet with a valid signature and valid public key but is not belonging to the group ?

Reply:
Above we `util.Concat(proposal.Remaining, proposal.Joining)` to find the public key of the person purporting to send the packet.

In principle, somebody outside the network could send a proposal with a valid signature, but their eligibility to be `Leader` is checked in the state machine.
Similarly, someone in the group could try and trigger execution (with a valid signature for themselves), but if they aren't leader this would also be rejected by the state machine.

Not in this PR, but in the [gossip one](https://github.com/drand/drand/pull/1230) I've added a metadata object to each packet binding the `Address` and the `Signature` and there is some extra checking in the state machine for other packets